src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -858,76 +858,103 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher)
-                finally
-                {
-                    if (!success && backoffAllClients)
-                    {
-                        _logger.LogWarning(LogHelper.BuildLastEndpointFailedMessage(previousEndpoint?.ToString()));
-
-                        do
-                        {
-                            _configClientManager.UpdateClientStatus(currentClient, success);
-                            clientEnumerator.MoveNext();
-                            currentClient = clientEnumerator.Current;
-                        }
-                        while (currentClient != null);
-                    }
-                    else
-                    {
-                        _configClientManager.UpdateClientStatus(currentClient, success);
-                    }
-                }
-
-                Uri currentEndpoint = _configClientManager.GetEndpointForClient(clientEnumerator.Current);
-
-                if (previousEndpoint != currentEndpoint)
-                {
-                    _logger.LogWarning(LogHelper.BuildFailoverMessage(previousEndpoint?.ToString(), currentEndpoint?.ToString()));
-                }
-
-                previousEndpoint = currentEndpoint;
-            }
-        }
+            IEnumerator<ConfigurationClient> clientEnumerator = null;
+
+            try
+            {
+                clientEnumerator = clients.GetEnumerator();
+                clientEnumerator.MoveNext();
+
+                Uri previousEndpoint = _configClientManager.GetEndpointForClient(clientEnumerator.Current);
+                ConfigurationClient currentClient;
+
+                bool autoDiscovered = false;
+
+                while (true)
+                {
+                    bool success = false;
+                    bool backoffAllClients = false;
+
+                    cancellationToken.ThrowIfCancellationRequested();
+                    currentClient = clientEnumerator.Current;
+
+                    try
+                    {
+                        T result = await funcToExecute(currentClient).ConfigureAwait(false);
+                        success = true;
+
+                        return result;
+                    }
+                    catch (Exception ex) when (ex is RequestFailedException ||
+                                               ex is AggregateException)
+                    {
+                        if (!IsFailOverable(ex))
+                        {
+                            backoffAllClients = true;
+                            throw;
+                        }
+                        else if (!clientEnumerator.MoveNext())
+                        {
+                            if (_options.AutoDiscoverReplica && !autoDiscovered)
+                            {
+                                IEnumerable<ConfigurationClient> autoFailoverClients = await _configClientManager.GetAutoFailoverClients(cancellationToken).ConfigureAwait(false);
+                                autoDiscovered = true;

```

Comment:
The provider defers to ConfigurationClientManager to provide available clients. I recommend we internalize all knowledge about obtaining clients in the manager and keep the logic here the same.

Reply:
Explained the consideration of the reason I put it here https://github.com/Azure/AppConfiguration-DotnetProvider/pull/442#discussion_r1338239727